pkg/cloud/services/ec2/ami.go

```diff
@@ -94,8 +103,43 @@ func GenerateAmiName(amiNameFormat, baseOS, kubernetesVersion string) (string, e
 	return templateBytes.String(), nil
 }
 
+// Determine architecture based on instance type.
+func (s *Service) pickArchitectureForInstanceType(instanceType string) (string, error) {
+	descInstanceTypeInput := &ec2.DescribeInstanceTypesInput{
+		InstanceTypes: []*string{&instanceType},
+	}
+	describeInstanceTypeResult, err := s.EC2Client.DescribeInstanceTypes(descInstanceTypeInput)
+	if err != nil {
+		return "", err
+	}
+
+	supportedArchs := describeInstanceTypeResult.InstanceTypes[0].ProcessorInfo.SupportedArchitectures
+
+	logger := s.scope.GetLogger().WithValues("instance type", instanceType, "supported architectures", supportedArchs)
+	logger.Info("Obtained a list of supported architectures for instance type")
+
+	// Loop over every supported architecture for the instance type
+	architecture := ""
+	for _, a := range supportedArchs {
```

Comment:
Once it's found we should `break`.

Reply:
I'm... not even sure this is correct, right? This looks weird. This means if ARM is the last thing, it will choose that. But if it happens to come before AMD then and might be selected last, right? This will select whatever comes last in the list.

Reply:
There was some previous conversation around this logic https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4054#pullrequestreview-1315802219

The value of `supportedArchs` is almost always a single architecture, e.g `amd64`. In some cases there's multiple architectures like in the case of some small `t2` instances, they have both `amd64` and `i386` - if we ever encounter that we only want to select `amd64` anyway. An instance should never be able to support both `amd64` and `arm64` so we shouldn't encounter any issues from doing the checks like this.

As mentioned in the previous comment, it means if AWS decide to add more architectures like `riscv64` we'll have to update the switch statement.